### PR TITLE
add server-side hook: proc-receive

### PR DESCRIPTION
**Pull request by Cactusinhand**
- [x] I provide my work under the [project license](https://github.com/progit/progit2/blob/main/LICENSE.asc).
- [x] I grant such license of my work as is required for the purposes of future print editions to [Ben Straub](https://github.com/ben) and [Scott Chacon](https://github.com/schacon).

## Changes

- Add new server side hook `proc-receive`

## Context
In git hooks, a new server side hook named `proc-receive` have been added in 2020. This is a very usefully hook, I think it shoud be known by more people, especialy the readers of this book.

### CODE REVIEW

**Comment by Cactusinhand** [comment]
Hi, @jiangxin I may need your input on this change, please correct me if I'm missing anything.

File: book/08-customizing-git/sections/hooks.asc [hunk 1]
```diff
@@ -111,6 +111,14 @@ The first script to run when handling a push from a client is `pre-receive`.
 It takes a list of references that are being pushed from stdin; if it exits non-zero, none of them are accepted.
 You can use this hook to do things like make sure none of the updated references are non-fast-forwards, or to do access control for all the refs and files they're modifying with the push.
 
+===== `proc-receive`
+
+The `proc-receive` script is quite different from other server-side scripts, it is invoked by git push process, but can change the workflow or process between `pre-reveive` and `post-receive` hook.
+This's because if the upcoming update references match the specified reference pattern, which can be configured by `receive.procReceiveRefs`, then `proc-receive` will take over the whole flow
+after `pre-receive`, then ignore the `update` script. It can decide to create, update or rename new reference, or even fallback to previous process.
```
Comment (jiangxin) [inline book/08-customizing-git/sections/hooks.asc:118]: "ignore the `update` script" is not the key point, we can ignore it here. We know that all the commands are fed to "pre-receive" hook in a batch, and each command to be executed by internal "exec_command()"  will be fed to the "update" hook, and the "proc-receive" hook should implement the "update" hook for other commands.

File: book/08-customizing-git/sections/hooks.asc [hunk 2]
```diff
@@ -111,6 +111,14 @@ The first script to run when handling a push from a client is `pre-receive`.
 It takes a list of references that are being pushed from stdin; if it exits non-zero, none of them are accepted.
 You can use this hook to do things like make sure none of the updated references are non-fast-forwards, or to do access control for all the refs and files they're modifying with the push.
 
+===== `proc-receive`
+
+The `proc-receive` script is quite different from other server-side scripts, it is invoked by git push process, but can change the worklow or process between `pre-reveive` and
+`post-receive` script. This's because if the upcoming update references match the specified reference pattern, which can be configured by `receive.procReceiveRefs`, then `proc-receive`
+will take over the whole flow after `pre-receive`. It can decide to create, update or rename new reference, or even fallback to previous process.
+In other word, it can outsource reference processing flow, this give a system administrator to customize very useful policy for pushing workflow, such as Gerrit-like workflow in Git.
```
Comment (Morganov) [inline book/08-customizing-git/sections/hooks.asc:119]:
```suggestion
In other words, it can outsource reference processing flow, this gives a system administrator to customize a very useful policy for pushing workflow, such as Gerrit-like workflow in Git.
```

Additionally, please, follow the rule: 1 sentence per 1 line.

**Review by Morganov — CHANGES_REQUESTED** [review]

File: book/08-customizing-git/sections/hooks.asc [hunk 3]
```diff
@@ -111,6 +111,14 @@ The first script to run when handling a push from a client is `pre-receive`.
 It takes a list of references that are being pushed from stdin; if it exits non-zero, none of them are accepted.
 You can use this hook to do things like make sure none of the updated references are non-fast-forwards, or to do access control for all the refs and files they're modifying with the push.
 
+===== `proc-receive`
+
+The `proc-receive` script is quite different from other server-side scripts, it is invoked by git push process, but can change the worklow or process between `pre-reveive` and `post-receive` script.
+This's because if the upcoming update references match the specified reference pattern, which can be configured by `receive.procReceiveRefs`, then `proc-receive` will take over the whole flow after `pre-receive`.
+It can decide to create, update or rename new reference, or even fallback to previous process.
+In other words, it can outsource reference processing flow, this gives a system administrator to customize a very useful policy for pushing workflow, such as Gerrit-like workflow in Git.
+This script runs once at push process, and can stop the push process.
```
Comment (ben) [inline book/08-customizing-git/sections/hooks.asc:120]: Some tweaks on content and tone:

```suggestion
The `proc-receive` script is quite different from other server-side scripts.
It is invoked by git push process, but can change the worklow or process between `pre-receive` and `post-receive` script.
If the incoming update references match the specified reference pattern (`receive.procReceiveRefs`), then `proc-receive` will take over the whole flow after `pre-receive`.
It can decide to create, update or rename new reference, or even fall back to previous process.

In other words, it can outsource reference processing flow.
This gives a system administrator the ability to customize a pushing workflow, like that used in a Gerritt project.
```

**Review by revmarcuspeter-cyber — APPROVED** [review]
